source/MaterialXCore/Document.cpp

```diff
@@ -152,16 +152,21 @@ void Document::initialize()
 }
 
 NodeDefPtr Document::addNodeDefFromGraph(const NodeGraphPtr nodeGraph, const string& nodeDefName, const string& category,
-                                         const string& version, bool isDefaultVersion, const string& group, string& newGraphName)
+                                         const string& version, bool isDefaultVersion, const string& group, 
+                                         string& newGraphName, const string& namespaceString)
 {
     if (!nodeGraph || nodeDefName.empty() || category.empty())
     {
         throw Exception("Invalid input arguments specified for nodedef creation");
     }
 
-    if (getNodeDef(nodeDefName))
+    // Always store nodedef name as fully qualified name
+    bool isNameSpaced = !namespaceString.empty();
+    string qualifiedNodeDefName = isNameSpaced ? (namespaceString + NAME_PREFIX_SEPARATOR + nodeDefName) : nodeDefName;
+    
+    if (getNodeDef(qualifiedNodeDefName))
```

Comment:
I think getNodeDef could be updated to have a namespace variant. So that every caller is not required to append namespace.  or a getQualifiedNodeDef() where namespace is required.  
I think this is going be become hard to clean up later.

Reply:
Actually getNodeDef relative to a node which has the ability to look for qualified and unqualified names.  
This is the general use case.

This is at the Document level which has no such contextual information and just looks up the identifier in the table of implementation associations. Here we always want to store a fully qualified name so I've handled this as part of this utility method.  Note that we also do then when we "load libraries" of definitions -- the nodedef identifiers are also stored fully qualified / namespaced.